### DISCUSSION TITLE
V4 (hardfork)

### DISCUSSION THREAD

**Discussion by kroggen** [aergoio/aergo#277] [General]
This is a place to discuss V4 and things that should be done before it

**Comment by kroggen** [comment]
## TODO

### Implement Tests

* **DONE:** The existing tests must be run for all the hardfork versions instead of just 1
* **DONE:** Implement tests listed on #241
* **DONE:** Use `t.Helper()` on `readLuaCode()` (check [here](https://stackoverflow.com/questions/71266376/show-original-source-line-when-wrapping-testing-t-errorf))
* **DONE:** Integration tests for `dpos` and `raft`
* **DONE:** Add tests for #193
* **DONE:** Add tests for #270
* **DONE:** Add tests for #177
* Replace most of `system.print()` with assertions

**Comment by kroggen** [comment]
## Things to do BEFORE the release of V4

* Fix the [Security issues](https://github.com/aergoio/aergo/security/dependabot)
* Apply #200 and update the db on all nodes
* Apply last LuaJIT fixes

**Reply by gosunuts** [reply]
#272, #274 sync test completed.

**Comment by gosunuts** [comment]
can you share change list in V4 hardfork? 
and, I plan to add gas limit on block after v4, to prevent indeterministic timeout.

**Reply by kroggen** [reply]
Those PRs marked as V4 may be included, probably all of them.

I don't know if Composable Transactions will be added on V4 or V5, but this decision can be made once all those other things are working properly
